File: Week_04/word_ladder/Solution.java
package word_ladder;

import java.util.*;

/**
 * BFS，单向搜索
 * Time: O(n*26*m*m), n=wordList.size, m = wordLength (1st m = iterating to change every char in word, 2nd m = creating string from char array)
 * Space: O(n)
 */
class Solution1 {
    public static void main(String[] args) {
        new Solution1().ladderLength("hit", "cog", Arrays.asList("hot", "dot", "dog", "lot", "log", "cog"));
    }

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (wordSet.size() == 0 || !wordSet.contains(endWord)) {
            return 0;
        }

        Queue<String> queue = new ArrayDeque<>();
        queue.add(beginWord);
        Set<String> visited = new HashSet<>();
        visited.add(beginWord);

        int step = 1;
        while (!queue.isEmpty()) {
            int curSize = queue.size();
            for (int i = 0; i < curSize; i++) {
                String curWord = queue.poll();
                if (changeLetterInWord(curWord, endWord, queue, visited, wordSet))
                    return step + 1;
            }
            step++;
        }
        return 0;
    }

    private boolean changeLetterInWord(String curWord, String endWord, Queue<String> queue, Set<String> visited, Set<String> wordSet) {
        char[] chars = curWord.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            char originalChar = chars[i];
            for (char c = 'a'; c <= 'z'; c++) {
                if (c == originalChar)
                    continue;
                chars[i] = c;
                String nextWord = new String(chars);
                if (wordSet.contains(nextWord)) {
                    if (nextWord.equals(endWord))
                        return true;
                    if (!visited.contains(nextWord)) {
                        queue.add(nextWord);
                        visited.add(nextWord);
                    }
                }
            }
            chars[i] = originalChar;
        }
        return false;
    }
}

/**
 * BFS 双向搜索
 */
class Solution2 {
    public static void main(String[] args) {
        int i = new Solution2().ladderLength("hit", "cog", Arrays.asList("hot", "dot", "dog", "lot", "log", "cog"));
        System.out.println(i);
    }

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (wordSet.size() == 0 || !wordSet.contains(endWord))
            return 0;

        Set<String> toVisitFromBegin = new HashSet<>();
        toVisitFromBegin.add(beginWord);

        Set<String> toVisitFromEnd = new HashSet<>();
        toVisitFromEnd.add(endWord);

        Set<String> visited = new HashSet<>();
        visited.add(beginWord);
        visited.add(endWord);

        int step = 1;
        while (!toVisitFromBegin.isEmpty() && !toVisitFromEnd.isEmpty()) {
            if (toVisitFromBegin.size() > toVisitFromEnd.size()) {
                Set<String> tmp = toVisitFromBegin;
                toVisitFromBegin = toVisitFromEnd;
                toVisitFromEnd = tmp;
            }

            Set<String> nextLevelToVisit = new HashSet<>();
            for (String curWord : toVisitFromBegin) {
                if (changeOneLetterInWord(curWord, toVisitFromEnd, visited, wordSet, nextLevelToVisit))
                    return step + 1;
            }
            step++;
            toVisitFromBegin = nextLevelToVisit;
        }
        return 0;
    }

    private boolean changeOneLetterInWord(String curWord, Set<String> toVisitFromEnd, Set<String> visited, Set<String> wordSet, Set<String> nextLevelToVisit) {
        char[] chars = curWord.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            char originalChar = chars[i];
            for (char c = 'a'; c <= 'z'; c++) {
                chars[i] = c;
                String nextWord = new String(chars);
                if (wordSet.contains(nextWord)) {
                    if (toVisitFromEnd.contains(nextWord))
                        return true;
                    if (!visited.contains(nextWord)) {
                        nextLevelToVisit.add(nextWord);
                        visited.add(nextWord);
                    }
                }
            }
            chars[i] = originalChar;
        }
        return false;
    }
}

File: Week_04/word_ladder/Solution.py
from typing import List

# 单向 BFS
class Solution1:
    def ladderLength(self, beginWord, endWord, wordList):
        wordList = set(wordList)
        queue = collections.deque([[beginWord, 1]])
        while queue:
            word, length = queue.popleft()
            if word == endWord:
                return length
            for i in range(len(word)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    next_word = word[:i] + c + word[i+1:]
                    if next_word in wordList:
                        wordList.remove(next_word)
                        queue.append([next_word, length + 1])
        return 0


# 双向 BFS
class Solution2:
    def ladderLength(self, begin_word: str, end_word: str, word_list: List[str]) -> int:
        word_set = set(word_list)
        if not word_set or end_word not in word_set:
            return 0

        begin_set, end_set, visited = set([begin_word]), set([end_word]), set([begin_word, end_word])

        step = 1
        while begin_set and end_set:
            if len(begin_set) > len(end_set):
                begin_set, end_set = end_set, begin_set

            next_to_visit = set()
            for word in begin_set:
                word_list = list(word)

                for i in range(len(word_list)):
                    old_char = word_list[i]

                    for char in 'abcdefghijklmnopqrstuvwxyz':
                        word_list[i] = char
                        new_word = ''.join(word_list)

                        if new_word in word_set:
                            if new_word in end_set:
                                return step + 1
                            if new_word not in visited:
                                next_to_visit.add(new_word)
                                visited.add(new_word)
                    word_list[i] = old_char
            step += 1
            begin_set = next_to_visit
        return 0


print(Solution().ladderLength('hit','cog',["hot","dot","dog","lot","log","cog"]))